--- linux-2.6.18.orig/drivers/md/raid5.c	2011-11-01 22:03:03.000000000 -0600
+++ linux-2.6.18/drivers/md/raid5.c	2011-11-01 22:04:03.000000000 -0600
@@ -2470,8 +2470,6 @@
 					locked++;
 					set_bit(R5_Wantwrite, &sh->dev[i].flags);
 				}
-			/* after a RECONSTRUCT_WRITE, the stripe MUST be in-sync */
-			set_bit(STRIPE_INSYNC, &sh->state);
 
 			if (test_and_clear_bit(STRIPE_PREREAD_ACTIVE, &sh->state)) {
 				atomic_dec(&conf->preread_active_stripes);
