Index: linux-2.6.32-131.0.15.el6.x86_64/drivers/md/raid5.c
===================================================================
--- linux-2.6.32-131.0.15.el6.x86_64.orig/drivers/md/raid5.c	2011-05-19 21:21:11.000000000 +0300
+++ linux-2.6.32-131.0.15.el6.x86_64/drivers/md/raid5.c	2011-05-19 21:23:31.000000000 +0300
@@ -5322,6 +5322,12 @@ static int run(mddev_t *mddev)
 		blk_queue_io_min(mddev->queue, chunk_size);
 		blk_queue_io_opt(mddev->queue, chunk_size *
 				 (conf->raid_disks - conf->max_degraded));
+		{
+		int stripe_size = conf->chunk_sectors * (conf->previous_raid_disks - conf->max_degraded);
+
+		blk_queue_max_hw_sectors(mddev->queue, stripe_size);
+		blk_queue_max_segments(mddev->queue, stripe_size);
+		}
 
 		list_for_each_entry(rdev, &mddev->disks, same_set)
 			disk_stack_limits(mddev->gendisk, rdev->bdev,
