--- linux-stage.orig/fs/ext4/ext4.h	2015-08-25 13:34:34.492894882 +0530
+++ linux-stage/fs/ext4/ext4.h	2015-08-25 13:23:08.211892985 +0530
@@ -1992,6 +1992,8 @@ extern int ext4_group_extend(struct supe
 				ext4_fsblk_t n_blocks_count);
 
 /* super.c */
+#define HAVE_EXT4_COMMIT_SUPER
+extern int ext4_commit_super(struct super_block *sb, int sync);
 extern void *ext4_kvmalloc(size_t size, gfp_t flags);
 extern void *ext4_kvzalloc(size_t size, gfp_t flags);
 extern void ext4_kvfree(void *ptr);
--- linux-stage.orig/fs/ext4/super.c	2015-08-25 13:34:34.491894938 +0530
+++ linux-stage/fs/ext4/super.c	2015-08-25 13:28:42.279894767 +0530
@@ -63,7 +63,6 @@ static struct ext4_features *ext4_feat;
 
 static int ext4_load_journal(struct super_block *, struct ext4_super_block *,
 			     unsigned long journal_devnum);
-static int ext4_commit_super(struct super_block *sb, int sync);
 static void ext4_mark_recovery_complete(struct super_block *sb,
 					struct ext4_super_block *es);
 static void ext4_clear_journal_err(struct super_block *sb,
@@ -4074,7 +4073,7 @@ static int ext4_load_journal(struct supe
 	return 0;
 }
 
-static int ext4_commit_super(struct super_block *sb, int sync)
+int ext4_commit_super(struct super_block *sb, int sync)
 {
 	struct ext4_super_block *es = EXT4_SB(sb)->s_es;
 	struct buffer_head *sbh = EXT4_SB(sb)->s_sbh;
@@ -4141,7 +4140,7 @@ static int ext4_commit_super(struct supe
 	}
 	return error;
 }
-
+EXPORT_SYMBOL(ext4_commit_super);
 /*
  * Have we just finished recovery?  If so, and if we are mounting (or
  * remounting) the filesystem readonly, then we will end up with a
