diff -rbBu linux-stage.orig/fs/ext4/super.c linux-stage/fs/ext4/super.c
--- linux-stage.orig/fs/ext4/super.c	2011-06-13 23:18:59.000000000 +0800
+++ linux-stage/fs/ext4/super.c	2011-06-23 03:06:58.000000000 +0800
@@ -2392,10 +2392,13 @@
 		upper_limit <<= blkbits;
 	}
 
-	/* 32-bit extent-start container, ee_block */
-	res = 1LL << 32;
+	/*
+	* 32-bit extent-start container, ee_block. We lower the maxbytes
+	* by one fs block, so ee_len can cover the extent of maximum file
+	* size
+	*/
+	res = (1LL << 32) - 1;
 	res <<= blkbits;
-	res -= 1;
 
 	/* Sanity check against vm- & vfs- imposed limits */
 	if (res > upper_limit)
