Index: linux-stage/fs/ext4/ext4.h
===================================================================
--- linux-stage.orig/fs/ext4/ext4.h
+++ linux-stage/fs/ext4/ext4.h
@@ -817,6 +817,14 @@ struct ext4_inode_info {
 	__u64 i_fs_version;
 };
 
+#ifndef O_32BITHASH
+# define O_32BITHASH	0x10000000
+#endif
+
+#ifndef O_64BITHASH
+# define O_64BITHASH	0x20000000
+#endif
+
 #define HAVE_DISK_INODE_VERSION
 
 /*
