diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 06328d3..51c9452 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -1849,6 +1849,10 @@ cleanup:
 	return err;
 }
 
+static int ext4_ext_fiemap_cb(struct inode *inode, struct ext4_ext_path *path,
+		       struct ext4_ext_cache *newex, struct ext4_extent *ex,
+		       void *data);
+
 int ext4_ext_walk_space(struct inode *inode, ext4_lblk_t block,
 			ext4_lblk_t num, ext_prepare_callback func,
 			void *cbdata)
@@ -1862,9 +1862,11 @@ int ext4_ext_walk_space(struct inode *inode, ext4_lblk_t block,
 	while (block < last && block != EXT_MAX_BLOCK) {
 		num = last - block;
 		/* find extent for this block */
-		down_read(&EXT4_I(inode)->i_data_sem);
+		if (func == &ext4_ext_fiemap_cb)
+			down_read(&EXT4_I(inode)->i_data_sem);
 		path = ext4_ext_find_extent(inode, block, path);
-		up_read(&EXT4_I(inode)->i_data_sem);
+		if (func == &ext4_ext_fiemap_cb)
+			up_read(&EXT4_I(inode)->i_data_sem);
 		if (IS_ERR(path)) {
 			err = PTR_ERR(path);
 			path = NULL;
